File: typescript/src/ancilis/cli/status.ts
/** ancilis status — developer's primary interaction point. */

import type { ResolvedConfig } from "../config/index.js";
import type { EvidenceSummary } from "../report/generator.js";
import { readFileSync, readdirSync } from "node:fs";
import { join } from "node:path";
import { sharedPathFrom } from "../shared-path.js";

const CONTROLS_DIR = sharedPathFrom(import.meta.url, "controls");

function normalizedDecisions(summary: EvidenceSummary): Record<string, number> {
  const normalized: Record<string, number> = {};
  for (const [key, value] of Object.entries(summary.decisions ?? {})) {
    normalized[key.trim().toUpperCase()] = value;
  }
  return normalized;
}

function loadControlDefs(): Map<string, Record<string, unknown>> {
  const controls = new Map<string, Record<string, unknown>>();
  try {
    const files = readdirSync(CONTROLS_DIR).filter(f => f.endsWith(".json")).sort();
    for (const file of files) {
      const data = JSON.parse(readFileSync(join(CONTROLS_DIR, file), "utf-8"));
      controls.set(data.id, data);
    }
  } catch { /* ok */ }
  return controls;
}

export function formatStatus(config: ResolvedConfig, summary: EvidenceSummary, verbose = false): string {
  const lines: string[] = [];
  const controlDefs = loadControlDefs();

  lines.push(`Ancilis — ${config.agentName}`);
  lines.push(`  Mode: ${config.mode}`);

  const enabled = [...config.controls.values()].filter(c => c.enabled);
  const controlStats = summary.control_pass_rates ?? {};

  // Honest per-control bucketing (mirrors Python status). The headline must
  // never report "all passing" while a control is failing, flagged, or has
  // not actually been evaluated (never evaluated, or only SKIP results).
  let verified = 0;
  let attested = 0;
  let pendingCount = 0;
  let flaggedCount = 0;
  let failingCount = 0;
  for (const cs of enabled) {
    const stats = controlStats[cs.controlId] ?? {};
    const cdef = controlDefs.get(cs.controlId) ?? {};
    const supportLevel = (cdef.support_level as string | undefined) ?? "runtime_evaluator";
    const failCount = (stats.FAIL ?? 0) + (stats.ERROR ?? 0);
    const flagCount = stats.FLAG ?? 0;
    const passCount = stats.PASS ?? 0;
    if (failCount > 0) {
      failingCount += 1;
    } else if (flagCount > 0) {
      flaggedCount += 1;
    } else if (passCount > 0) {
      if (supportLevel === "attestation") attested += 1;
      else verified += 1;
    } else {
      // No PASS/FAIL/FLAG: never evaluated, or only SKIP (e.g. an
      // attestation control awaiting its first attestation).
      pendingCount += 1;
    }
  }

  const totalEvals = summary.total_evaluations;
  let controlSuffix: string;
  if (totalEvals === 0) {
    controlSuffix = "not yet evaluated";
  } else {
    const parts: string[] = [];
    if (verified) parts.push(`${verified} runtime-verified`);
    if (attested) parts.push(`${attested} attestation-passing`);
    if (pendingCount) parts.push(`${pendingCount} pending`);
    if (flaggedCount) parts.push(`${flaggedCount} flagged`);
    if (failingCount) parts.push(`${failingCount} failing`);
    controlSuffix = parts.length > 0 ? parts.join(", ") : "not yet evaluated";
    // Only an unqualified clean state earns the "all passing" affirmation.
    if (!pendingCount && !flaggedCount && !failingCount && (verified || attested)) {
      controlSuffix += " — all passing";
    }
  }
  lines.push(`  Controls: ${enabled.length} active, ${controlSuffix}`);

  // Certification one-liners
  for (const certId of config.activeCertifications) {
    lines.push(`  ${certId.toUpperCase()}: active`);
  }

  // Overlay one-liners
  for (const [, oa] of [...config.activeOverlays.entries()].sort((a, b) => a[0].localeCompare(b[0]))) {
    let trigger = "";
    if (oa.triggeredBy.length > 0) {
      const first = oa.triggeredBy[0];
      if (first !== undefined && first.includes(" via ")) {
        trigger = ` — triggered by ${first.split(" via ")[1]} declaration`;
      }
    }
    lines.push(`  ${oa.name}: active${trigger}`);
  }

  // Evaluation counts
  const total = summary.total_evaluations;
  const blocked = normalizedDecisions(summary).BLOCK ?? 0;
  if (total > 0) {
    lines.push(`  Tool calls: ${total.toLocaleString()} evaluated, ${blocked} blocked`);
  } else {
    lines.push("  No evaluations recorded yet. Middleware is collecting data.");
  }

  // Verbose detail
  if (verbose) {
    lines.push("");
    lines.push("  Controls:");
    for (const cs of [...enabled].sort((a, b) => a.controlId.localeCompare(b.controlId))) {
      const cdef = controlDefs.get(cs.controlId) ?? {};
      const displayName = (cdef.display_name as string) ?? cs.name;
      const stats = controlStats[cs.controlId] ?? {};
      const failCount = (stats.FAIL ?? 0) + (stats.ERROR ?? 0);
      const flagCount = stats.FLAG ?? 0;

      const passCount = stats.PASS ?? 0;
      const totalCount = Object.values(stats).reduce((a, b) => a + b, 0);

      let mark: string, statusStr: string;
      if (totalCount === 0) {
        mark = "\u2013";
        statusStr = "not yet evaluated";
      } else if (failCount > 0) {
        mark = "\u2717";
        statusStr = `failing (${failCount} failures)`;
      } else if (flagCount > 0) {
        // A flag is a deviation for review \u2014 not a pass.
        mark = "!";
        statusStr = `flagged (${flagCount} flag${flagCount !== 1 ? "s" : ""})`;
      } else if (passCount > 0) {
        mark = "\u2713";
        statusStr = "passing";
      } else {
        // Evaluated only as SKIP (e.g. an attestation control not yet
        // attested) \u2014 pending, not passing.
        mark = "\u25cb";
        statusStr = "pending (attestation required)";
      }
      lines.push(`    ${mark} ${displayName} — ${statusStr}`);
    }

    if (config.activeCertifications.length > 0 || config.activeOverlays.size > 0) {
      lines.push("");
      lines.push("  Activation:");
      for (const certId of config.activeCertifications) {
        lines.push(`    ${certId.toUpperCase()} certification active — ${enabled.length} controls enforcing`);
      }
      for (const [, oa] of [...config.activeOverlays.entries()].sort((a, b) => a[0].localeCompare(b[0]))) {
        let trigger = "";
        if (oa.triggeredBy.length > 0) {
          const first = oa.triggeredBy[0];
          if (first !== undefined && first.includes(" via ")) {
            trigger = ` — triggered by ${first.split(" via ")[1]} declaration`;
          }
        }
        lines.push(`    ${oa.name} overlay active${trigger}`);
      }
    }

    if (total > 0) {
      const chainStatus = summary.chain_valid ? "intact" : "BROKEN";
      lines.push(`  Evidence records: ${total.toLocaleString()} stored, hash chain ${chainStatus}`);
    }
  }

  return lines.join("\n");
}

File: python/src/ancilis/cli/status.py
"""ancilis status — developer's primary interaction point."""

from __future__ import annotations

from pathlib import Path
from typing import Any

import click

from ancilis.config import ResolvedConfig, load_config, load_control_definitions
from ancilis.evidence.store import EvidenceStore


def _normalized_decisions(summary: dict[str, Any]) -> dict[str, int]:
    decisions = summary.get("decisions", {})
    normalized: dict[str, int] = {}
    for key, value in decisions.items():
        normalized[str(key).strip().upper()] = int(value)
    return normalized


def _load_config_safe(config_path: str | None) -> ResolvedConfig | None:
    try:
        if config_path:
            return load_config(path=config_path)
        return load_config()
    except (FileNotFoundError, ValueError) as e:
        click.echo(f"Error loading config: {e}", err=True)
        return None


def _format_status(
    config: ResolvedConfig,
    evidence: EvidenceStore,
    verbose: bool,
    session_id: str | None = None,
) -> str:
    lines: list[str] = []
    control_defs = load_control_definitions()

    lines.append(f"Ancilis — {config.agent_name}")
    lines.append(f"  Mode: {config.mode}")

    # Control summary
    enabled = [c for c in config.controls.values() if c.enabled]
    summary = evidence.get_summary(session_id=session_id)
    control_stats = summary.get("control_pass_rates", {})

    all_passing = True
    warnings: list[dict[str, str]] = []
    for cs in enabled:
        stats = control_stats.get(cs.control_id, {})
        if stats.get("FAIL", 0) > 0 or stats.get("ERROR", 0) > 0:
            all_passing = False
        # Collect FLAG results as warnings
        if stats.get("FLAG", 0) > 0:
            cdef = control_defs.get(cs.control_id, {})
            display = cdef.get("display_name", cs.name).lower()
            warnings.append({
                "category": display,
                "message": f"{cdef.get('display_name', cs.name)} flagged deviations",
                "hint": "Review recent activity: ancilis report --period 1d",
            })

    total = summary.get("total_evaluations", 0)
    if total == 0:
        passing_str = "not yet evaluated"
    elif all_passing:
        passing_str = "all passing"
    else:
        passing_str = "issues detected"
    lines.append(f"  Controls: {len(enabled)} active, {passing_str}")

    # Certification one-liners
    if config.active_certifications:
        for cert_id in config.active_certifications:
            lines.append(f"  {cert_id.upper()}: active")

    # Overlay one-liners
    if config.active_overlays:
        for _oid, oa in sorted(config.active_overlays.items()):
            trigger = ""
            if oa.triggered_by:
                first = oa.triggered_by[0]
                if " via " in first:
                    data_type = first.split(" via ")[1]
                    trigger = f" — triggered by {data_type} declaration"
            lines.append(f"  {oa.name}: active{trigger}")
    decisions = _normalized_decisions(summary)
    blocked = decisions.get("BLOCK", 0)
    if total > 0:
        lines.append(f"  Tool calls: {total:,} evaluated, {blocked} blocked")
    else:
        lines.append("  No evaluations recorded yet. Run your agent with Ancilis to start collecting evidence.")

    # Verbose: per-control detail
    if verbose:
        lines.append("")
        lines.append("  Controls:")
        for cs in sorted(enabled, key=lambda c: c.control_id):
            cdef = control_defs.get(cs.control_id, {})
            display_name = cdef.get("display_name", cs.name)
            stats = control_stats.get(cs.control_id, {})
            fail_count = stats.get("FAIL", 0) + stats.get("ERROR", 0)
            flag_count = stats.get("FLAG", 0)

            total_evals = sum(stats.values()) if stats else 0
            if total_evals == 0:
                mark = "–"
                status_str = "not yet evaluated"
            elif fail_count > 0:
                mark = "\u2717"
                status_str = f"failing ({fail_count} failures)"
            elif flag_count > 0:
                mark = "\u2713"
                status_str = f"passing ({flag_count} flags)"
            else:
                mark = "\u2713"
                status_str = "passing"
            lines.append(f"    {mark} {display_name} — {status_str}")

        # Activation details
        if config.active_certifications or config.active_overlays:
            lines.append("")
            lines.append("  Activation:")
            for cert_id in config.active_certifications:
                count = len(enabled)
                lines.append(f"    {cert_id.upper()} certification active — {count} controls enforcing")
            for _oid, oa in sorted(config.active_overlays.items()):
                trigger = ""
                if oa.triggered_by:
                    first = oa.triggered_by[0]
                    if " via " in first:
                        data_type = first.split(" via ")[1]
                        trigger = f" — triggered by {data_type} declaration"
                lines.append(f"    {oa.name} overlay active{trigger}")

        # Evidence summary
        if total > 0:
            chain_valid = summary.get("chain_valid", True)
            chain_status = "intact" if chain_valid else "BROKEN"
            lines.append(f"  Evidence records: {total:,} stored, hash chain {chain_status}")

    # Warnings (always shown)
    if warnings:
        lines.append("")
        lines.append("  Warnings:")
        for w in warnings:
            lines.append(f"    [{w['category']}] {w['message']}")
            lines.append(f"            {w['hint']}")

    return "\n".join(lines)


@click.command()
@click.option("--verbose", "-v", is_flag=True, help="Show detailed status with per-control breakdown")
@click.option("--config", "config_path", default=None, help="Path to ancilis.yaml")
@click.option("--db", "db_path", default=None, help="Path to evidence database")
@click.option("--session", "session_id", default=None, help="Scope to a specific session ID")
def status(verbose: bool, config_path: str | None, db_path: str | None, session_id: str | None) -> None:
    """Show current agent security posture."""
    config = _load_config_safe(config_path)
    if config is None:
        raise SystemExit(1)

    store = EvidenceStore(config, db_path=db_path)
    try:
        output = _format_status(config, store, verbose, session_id=session_id)
        click.echo(output)
    finally:
        store.close()
